src/include/gcc.h

```diff
@@ -89,38 +89,29 @@
  * swap) operations.
  */
 
-#ifdef __clang__
-/*
- * We avoid __sync_bool_compare_and_swap with due to problems with optimization
- * with some versions of clang. See http://llvm.org/bugs/show_bug.cgi?id=21499
- * for details.
- */
-#define	WT_ATOMIC_CAS(ptr, old, new)					\
-	(__sync_val_compare_and_swap(ptr, old, new) == (old))
-#else
-#define	WT_ATOMIC_CAS(ptr, old, new)					\
-	__sync_bool_compare_and_swap(ptr, old, new)
-#endif
+#define	WT_ATOMIC_CAS(ptr, oldp, new)					\
+	__atomic_compare_exchange_n(					\
+	    ptr, oldp, new, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST)
 #define	WT_ATOMIC_CAS_FUNC(name, vp_arg, old_arg, new_arg)		\
 static inline bool							\
 __wt_atomic_cas##name(vp_arg, old_arg, new_arg)				\
 {									\
-	return (WT_ATOMIC_CAS(vp, old, new));				\
+	return (WT_ATOMIC_CAS(vp, &old, new));				\
 }
 WT_ATOMIC_CAS_FUNC(8, uint8_t *vp, uint8_t old, uint8_t new)
 WT_ATOMIC_CAS_FUNC(16, uint16_t *vp, uint16_t old, uint16_t new)
 WT_ATOMIC_CAS_FUNC(32, uint32_t *vp, uint32_t old, uint32_t new)
 WT_ATOMIC_CAS_FUNC(v32,							\
-    volatile uint32_t *vp, volatile uint32_t old, volatile uint32_t new)
+    volatile uint32_t *vp, uint32_t old, volatile uint32_t new)
```

Comment:
The "overloads" for these `__atomic` functions don't support a pointer to a volatile type for the expected argument. I understand why you'd need to preserve the volatile specifier for the value that is getting written to but I think it's fine to remove it for the expected value.

Reply:
I agree, I think this change is correct.